### PR TITLE
Recsys functional

### PR DESCRIPTION
**Pull request by zkid18**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

File: catalyst/metrics/avg_precision.py [hunk 1]
```diff
@@ -45,17 +47,14 @@ def avg_precision(
...
     """
     k = min(outputs.size(1), k)
-    _, indices_for_sort = outputs.sort(descending=True, dim=-1)
-    targets_sorted_by_outputs = torch.gather(
-        targets, dim=-1, index=indices_for_sort
-    )
+    targets_sorted_by_outputs_at_k = process_recsys(outputs, targets, k)
 
-    targets_sorted_by_outputs_topk = targets_sorted_by_outputs[:, :k]
+    targets_sorted_by_outputs_topk = targets_sorted_by_outputs_at_k[:, :k]
```
Comment (zkid18) [inline catalyst/metrics/avg_precision.py]: to remove

File: catalyst/metrics/functional.py [hunk 2]
```diff
@@ -338,6 +366,37 @@ def get_default_topk_args(num_classes: int) -> Sequence[int]:
...
+    TO-DO: write tests for the mehtod.
+    """
+    lengths = [X.size(0) * X.size(1)for X in tensors]
+    uniques = np.unique(lengths)
+    if len(uniques) > 1:
+        raise ValueError("Found input variables with inconsistent numbers of"
+                         " samples:")
+
+
+def get_top_k(recsys_metric_at_k_gen: Generator) -> Tuple[float]:
```
Comment (zkid18) [inline catalyst/metrics/functional.py]: follow the `wrap_topk_metric2dict`  style

File: catalyst/metrics/functional.py [hunk 3]
```diff
@@ -338,6 +366,37 @@ def get_default_topk_args(num_classes: int) -> Sequence[int]:
     return result
 
 
+def check_consistent_length(*tesors):
```
Comment (zkid18) [inline catalyst/metrics/functional.py]: fix typo `tensors`

File: catalyst/metrics/functional.py [hunk 4]
```diff
@@ -81,6 +81,34 @@ def process_multiclass_components(
     return outputs, targets, num_classes
 
 
+def process_recsys(
```
Comment (zkid18) [inline catalyst/metrics/functional.py]: naming: `process_recsys_components`

File: catalyst/metrics/functional.py [hunk 5]
```diff
@@ -338,6 +367,22 @@ def get_default_topk_args(num_classes: int) -> Sequence[int]:
     return result
 
 
+def check_consistent_length(*tensors):
+    """Check that all arrays have consistent first dimensions.
```
Comment (Scitator) [inline catalyst/metrics/functional.py:372]: looks like we need another doc format :)

**Review by Scitator — COMMENTED** [review]
Let's firstly compute all statistic and then make 
```
for k in topk:
   ...
```
like [accuracy](https://github.com/catalyst-team/catalyst/blob/master/catalyst/metrics/accuracy.py#L19)

File: catalyst/metrics/functional.py [hunk 6]
```diff
@@ -421,6 +469,7 @@ def topk_metric_with_dict_output(*args, **kwargs):
 
 __all__ = [
     "process_multilabel_components",
+    "process_recsys_components",
     "get_binary_statistics",
     "get_multiclass_statistics",
     "get_multilabel_statistics",
```
Comment (Scitator) [inline catalyst/metrics/functional.py:474]: add `check_consistent_length`?

File: catalyst/metrics/ndcg.py [hunk 7]
```diff
@@ -28,55 +27,52 @@ def dcg(
         gain_function:
             String indicates the gain function for the ground truth labels.
             Two options available:
-            - `pow_rank`: torch.pow(2, x) - 1
+            - `exp_rank`: torch.pow(2, x) - 1
             - `rank`: x
```
Comment (Scitator) [inline catalyst/metrics/ndcg.py]: `linear_rank`?

**Review by Scitator — DISMISSED** [review]

**Comment by Scitator** [comment]
Thank you for the PR, I will handle transformers test 🚀

**Pull request merged**